### PR TITLE
Add caching support

### LINKED ISSUE #19: Leverage `wp_cache`
Because everyone loves cache.

### PR DESCRIPTION
**Pull request by mAAdhaTTah**
Fixes #19.

Gotta say, I'm kinda proud of this. I think this is a neat little implementation.

Still haven't implemented caching the queries though. I'm not sure how/where to bust those caches, but this will cut down on api calls.

### CODE REVIEW

**Comment by benbalter** [comment]
Nice work. What do you think about using WordPress's native `wp_cache` to take advantage of things like memcache or apc if present?

**Comment by mAAdhaTTah** [comment]
*tl;dr: The data doesn't change so let's just save it permanently.*

That was actually my first inclination, but a friend of mine point this out ([source](https://codex.wordpress.org/Class_Reference/WP_Object_Cache)):

> By default, the object cache is non-persistent. This means that data stored in the cache resides in memory only and only for the duration of the request. Cached data will not be stored persistently across page loads unless you install a persistent caching plugin.

So not going to be helpful if we're looking to save API calls across requests.

I also considered using Transients, which will cache to a persistent cache if it exists and to the options table if it doesn't. The reason I passed on that is you can lose the persistent cache, e.g. if there's a server restart, so you can't assume any data you stick in there is going to be around later.

Which isn't necessarily a big deal because we can always retrieve the data later via an API call if we wanted (same way you would rebuild a DB cache with new queries). However, because the data we're caching is immutable (blobs, trees, & commits will never change by sha), he suggested it made more sense to persist it permanently. Because it's structured so there's only 2 queries (once to load it up the first time it's used, and once to save everything before shutdown), it might actually be more performant than hitting the cache each time (though I haven't confirmed that specifically).

**Comment by benbalter** [comment]
Well reasoned. Makes total sense. Nice work. :+1:
